runners/direct-java/src/main/java/org/apache/beam/runners/direct/SplittableProcessElementsEvaluatorFactory.java

```diff
@@ -0,0 +1,191 @@
+    @SuppressWarnings({"unchecked", "rawtypes"})
+    TransformEvaluator<T> evaluator =
+        (TransformEvaluator<T>)
+            createEvaluator((AppliedPTransform) application, (CommittedBundle) inputBundle);
+    return evaluator;
+  }
+
+  @Override
+  public void cleanup() throws Exception {
+    DoFnLifecycleManagers.removeAllFromManagers(fnClones.asMap().values());
+  }
+
+  @SuppressWarnings({"unchecked", "rawtypes"})
+  private TransformEvaluator<KeyedWorkItem<String, ElementAndRestriction<InputT, RestrictionT>>>
+      createEvaluator(
+          AppliedPTransform<
+                  PCollection<KeyedWorkItem<String, ElementAndRestriction<InputT, RestrictionT>>>,
+                  PCollectionTuple, SplittableParDo.ProcessElements<InputT, OutputT, RestrictionT>>
+              application,
+          CommittedBundle<InputT> inputBundle)
+          throws Exception {
+    PCollection<KeyedWorkItem<String, ElementAndRestriction<InputT, RestrictionT>>> input =
+        application.getInput();
+
+    final SplittableParDo.ProcessElements<InputT, OutputT, RestrictionT> transform =
+        application.getTransform();
+
+    ElementAndRestrictionCoder<InputT, RestrictionT> elementAndRestrictionCoder =
+        ((ElementAndRestrictionCoder<InputT, RestrictionT>)
+            ((KeyedWorkItemCoder<String, ElementAndRestriction<InputT, RestrictionT>>)
+                    input.getCoder())
+                .getElementCoder());
+
+    // Cache the underlying DoFn rather than the ProcessFn itself.
+    DoFnLifecycleManager fnManager = fnClones.getUnchecked(transform.getFn());
+
+    try {
+      SplittableParDo.ProcessFn<InputT, OutputT, RestrictionT, ?> processFn =
+          new SplittableParDo.ProcessFn<>(
+              ((DoFn<InputT, OutputT>) fnManager.get()),
+              elementAndRestrictionCoder.getElementCoder(),
+              elementAndRestrictionCoder.getRestrictionCoder(),
+              input.getWindowingStrategy().getWindowFn().windowCoder());
+
+      processFn.setup();
+
+      String stepName = evaluationContext.getStepName(application);
+      final DirectExecutionContext.DirectStepContext stepContext =
+          evaluationContext
+              .getExecutionContext(application, inputBundle.getKey())
+              .getOrCreateStepContext(stepName, stepName);
+
+      ParDoEvaluator<KeyedWorkItem<String, ElementAndRestriction<InputT, RestrictionT>>, OutputT>
+          parDoEvaluator =
+              ParDoEvaluator.create(
+                  evaluationContext,
+                  stepContext,
+                  application,
+                  application.getInput().getWindowingStrategy(),
+                  processFn,
+                  transform.getSideInputs(),
+                  transform.getMainOutputTag(),
+                  transform.getSideOutputTags().getAll(),
+                  application.getOutput().getAll());
+
+      processFn.setStateInternalsFactory(
```

Comment:
The evaluator, such as `ParDoEvaluator`, is one per bundle. So `SplittableProcessElementsEvaluatorFactory.createEvaluator` is safe to pass them right here.